### PR TITLE
ci: Automate generation and update of docs-builder image

### LINKED ISSUE #24019: Automate creation and publication of docs-builder image for CI documentation workflow
We want to automate the creation and publication of `docs-builder` images, used in the CI workflow for docs, when updating the components of the documentation build system. Joe says:

> Seems like there's a similar issue with the docs-builder here as we have with the cilium-builder / cilium-runtime images. I agree for now it's OK to just handle this manually. If this bothers you, I would suggest looking a bit closer into (a) the way that the `images/` infrastructure detects changes and figures out canonical shas for a given set of input files, (b) the bot that can automatically build an image and update the commits in-line. Andre and I can help with any necessary docker repository configurations etc.

_Originally posted by @joestringer in https://github.com/cilium/cilium/pull/24014#pullrequestreview-1313867675_

### PR DESCRIPTION
**Pull request by qmonnet**
The docs-builder image is used for the CI wofkflow for testing documentation. Each time we change the list of requirements (Sphinx dependencies etc.) or the Dockerfile for that image, we have to generate a new image to use in the CI workflow.

So far these images have been generated and pushed to Docker manually (thanks Joe!). But this is not fun or practical [citation needed]. This commit introduces a new CI workflow to generate and push a new image when relevant, and to update the documentation workflow with the new tag and digest.

For Pull Requests from branches of the cilium/cilium repository, we can add a new commit directly to update the image. For branches from external forks, we require an additional step from the PR author to update the image reference.

Fixes: #24019

Status: Currently a draft, because we're missing scripts and docs regarding the update for PRs with branches on external forks. Also, nothing has been tested yet.

### CODE REVIEW

File: .github/workflows/build-images-docs-builder.yaml
Comment (joestringer) [inline .github/workflows/build-images-docs-builder.yaml]: TODO(@joestringer): Make sure these are configured and have the right level of access rights.

File: .github/workflows/build-images-docs-builder.yaml
Comment (joestringer) [inline .github/workflows/build-images-docs-builder.yaml]: cilium-runtime?

**Comment by joestringer** [comment]
Failed to log into dockerhub. I'm not sure whether we're going to be able to get the level of flexibility around bot write access to docker repos with DockerHub. Probably the next steps are then to switch to Quay.io instead:
- [ ] Initialize a docs-builder repo on quay.io/cilium
- [ ] Create new bot credentials for accessing that repo
- [ ] Create a new environment with these credentials
- [ ] Update this PR to change references for docs-builder to the new repo location
- [ ] Update this PR to reference the new environment + credential names

(cc @aanm)

File: .github/workflows/build-images-docs-builder.yaml
Comment (aanm) [inline .github/workflows/build-images-docs-builder.yaml]: If we are ok with this, this will only work for PRs opened from a branch in this repository.

Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml]: This is part of the second commit:
image

I intend to remove it after I manage to test the workflow successfully

**Comment by joestringer** [comment]
I've created https://quay.io/repository/cilium/docs-builder. If you configure the environment for the workflow for `docs-builder` and use the credentials from `QUAY_DOCS_BUILDER_USERNAME` and `QUAY_DOCS_BUILDER_PASSWORD`, it should be possible to push docs-builder images into that repository from here (via the environment, which requires @cilium/docs-structure approval to run the workflow).

**Comment by qmonnet** [comment]
<details><summary>Incremental diff</summary>

```diff
diff --git a/.github/workflows/build-images-docs-builder.yaml b/.github/workflows/build-images-docs-builder.yaml
index 01d1a7525dc5..bab24db224af 100644
--- a/.github/workflows/build-images-docs-builder.yaml
+++ b/.github/workflows/build-images-docs-builder.yaml
@@ -23,11 +23,11 @@ concurrency:
 jobs:
   build-and-push:
     timeout-minutes: 45
-    environment: release-base-images # To get access to AUTO_COMMITTER_*
+    environment: docs-builder
     runs-on: ubuntu-20.04
     steps:
-      - name: Checkout master branch to access local actions
-        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: Checkout main branch to access local actions
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
         with:
           ref: ${{ github.event.repository.default_branch }}
           persist-credentials: false
@@ -35,11 +35,11 @@ jobs:
       - name: Set environment variables
         uses: ./.github/actions/set-env-variables
 
-      - name: Set up docker buildx
-        uses: docker/setup-buildx-action@f03ac48505955848960e80bbb68046aa35c7b9e7 # v2.4.1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@4c0219f9ac95b02789c1075625400b2acbff50b1 # v2.9.1
 
       - name: Checkout source code
-        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
         with:
           persist-credentials: false
           ref: ${{ github.event.pull_request.head.sha }}
@@ -58,22 +58,23 @@ jobs:
         id: docs-builder-tag-in-repositories
         shell: bash
         run: |
-          if docker buildx imagetools inspect docker://${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }} &>/dev/null; then
+          if docker buildx imagetools inspect quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }} &>/dev/null; then
             echo exists="true" >> $GITHUB_OUTPUT
           else
             echo exists="false" >> $GITHUB_OUTPUT
           fi
 
-      - name: Login to DockerHub
+      - name: Login to quay.io
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
-        uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a # v2.1.0
+        uses: docker/login-action@465a07811f14bebb1938fbed4728c6a1ff8901fc # v2.2.0
         with:
-          username: ${{ secrets.DOCKER_HUB_USERNAME }}
-          password: ${{ secrets.DOCKER_HUB_PASSWORD }}
+          registry: quay.io
+          username: ${{ secrets.QUAY_DOCS_BUILDER_USERNAME }}
+          password: ${{ secrets.QUAY_DOCS_BUILDER_PASSWORD }}
 
       - name: Build docs-builder image
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
-        uses: docker/build-push-action@3b5e8027fcad23fda98b2e3ac259d8d67585f671 # v4.0.0
+        uses: docker/build-push-action@2eb1c1961a95fc15694676618e422e8ba1d63825 # v4.1.1
         id: docker_build_docs_builder
         with:
           provenance: false
@@ -81,7 +82,7 @@ jobs:
           file: ./Documentation/Dockerfile
           push: true
           tags: |
-            docker://${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}
+            quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}
 
       - name: Retrieve image digest
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
@@ -90,7 +91,7 @@ jobs:
           mkdir -p image-digest/
           echo "## docs-builder" > image-digest/docs-builder.txt
           echo "" >> image-digest/docs-builder.txt
-          echo "\`docker://${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}\`" >> image-digest/docs-builder.txt
+          echo "\`quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}\`" >> image-digest/docs-builder.txt
           echo "" >> image-digest/docs-builder.txt
 
       - name: Upload artifact digests
@@ -104,22 +105,24 @@ jobs:
       - name: Update docs-builder image reference in CI workflow
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         run: |
-          Documentation/update-docs-builder-image.sh "docker://${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}"
+          Documentation/update-docs-builder-image.sh "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}"
           git commit -sam "ci: update docs-builder"
 
       - name: Get token
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         id: get_token
-        uses: cilium/actions-app-token@350a14155dd9be79227f081310f4d77fdf792e76 # v0.21.0
+        uses: cilium/actions-app-token@61a6271ce92ba02f49bf81c755685d59fb25a59a # v0.21.1
         with:
           APP_PEM: ${{ secrets.AUTO_COMMITTER_PEM }}
           APP_ID: ${{ secrets.AUTO_COMMITTER_APP_ID }}
 
       - name: Push changes into PR
+        env:
+          REF: ${{ github.event.pull_request.head.ref }}
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         run: |
           git diff HEAD^
-          git push https://x-access-token:${{ steps.get_token.outputs.app_token }}@github.com/${{ env.QUAY_ORGANIZATION }}/cilium.git HEAD:${{ github.event.pull_request.head.ref }}
+          git push https://x-access-token:${{ steps.get_token.outputs.app_token }}@github.com/${{ env.QUAY_ORGANIZATION }}/cilium.git HEAD:"$REF"
 
   image-digests:
     name: Display digests
```
</details>

**Comment by qmonnet** [comment]
Looks like I've got something wrong

image

**Comment by qmonnet** [comment]
@joestringer Can you take a look, please? I can't find any obvious mistake in my YAML. Maybe I've got something wrong anyway, or maybe it has to do with the setup on either GH or quay.io? Let me know if there's any way I can help.

**Comment by joestringer** [comment]
I updated the username and password again, maybe last time I got them around the wrong way or something. After re-triggering the workflow it appears to log in this time :tada:

**Comment by qmonnet** [comment]
Thanks a lot!

The update script failed to find the image in the workflow, because we still reference the Docker image in the documentation workflow description. I pushed a preparatory commit to change that.

Now the deployment is stuck on the PR update:

image

@joestringer Am I right in supposing that `secrets.AUTO_COMMITTER_PEM` and `secrets.AUTO_COMMITTER_APP_ID` are not available yet from the `docs-builder` environment?

**Comment by joestringer** [comment]
That's right. @aanm are we able to add those credentials to the `docs-builder` environment as well?

**Review by aanm — COMMENTED** [review]
cc @qmonnet the secrets are now added

**Comment by qmonnet** [comment]
Looks good this time! The auto-committer bot added [a new commit](https://github.com/cilium/cilium/pull/24121/commits/18d022bdae53eb5a8423e59c1ea3359b5f06d080) to update the image, including the recent requirements changed from Dependabot (such robot team work).

image

The new commit re-triggered a deployment request and I accepted it to make sure it doesn't re-update (which didn't happen, so all good).

I'm going to remove the temporary commit I added to test the workflow on `pull_request`.

File: .github/workflows/build-images-docs-builder.yaml
Comment (joestringer) [inline .github/workflows/build-images-docs-builder.yaml:83]: Is there a way to edit this Dockerfile to expose environment variables from the host and echo them to the GHA output?

Comment (joestringer) [reply .github/workflows/build-images-docs-builder.yaml:83]: I guess with `environment: docs-builder` this requires docs-structure review anyway so as long as docs reviewers are aware that they need to check for env var abuse in any changes that are proposed against these files, then that is fine. Otherwise if docs-structure reviewers can't be trusted to enforce this then we'd need to think about how to ensure that folks with the appropriate knowledge are put in the path for running these checks.

Comment (joestringer) [reply .github/workflows/build-images-docs-builder.yaml:83]: I also wasn't 100% sure whether the secrets are available if a PR is opened from an external fork. If it requires push access to cilium/cilium then that is also a mitigating factor, as that requires org-member privileges IIRC.

**Review by joestringer — COMMENTED** [review]
Just one question, but I'll encourage additional scrutiny from other @cilium/github-sec members given the use of `pull_request_target` here.

File: .github/workflows/build-images-docs-builder.yaml
Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml:83]: > Is there a way to edit this Dockerfile to expose environment variables from the host and echo them to the GHA output?

Excellent question. As far as I'm aware, the only way would be to use the `ARG` directive in the Dockerfile, _**and**_ the `--build-arg` argument for the `docker` invocation; but the workflow file does not use the relevant `build-args` parameter for the `docker/build-push-action` action, so it shouldn't be possible to do it this way.

This being said, I'm not well versed on Docker security issues, so there could be something else I'm not aware of.

**Comment by qmonnet** [comment]
Rebasing to make CI happy

File: .github/workflows/build-images-docs-builder.yaml
Comment (nbusseneau) [reply .github/workflows/build-images-docs-builder.yaml:83]: This is correct. FYI we have the exact same setup in the image build workflows. These workflows run on `pull_request_target` so they run in a privileged context, and are indeed able to access repository secrets (and in general they have various dangerous permissions), so we must be careful as to what comes from the PR and how we use it.

In the case of `docker/build-push-action` using a Dockerfile coming from the PR it is fine because as said by Quentin, for the Dockerfile to be able to access the environment variables from the privileged workflow we would need to pass the secrets explicitly to the Dockerfile (as is done here: https://docs.docker.com/build/ci/github-actions/secrets/).

**Review by nbusseneau — APPROVED** [review]

**Review by joestringer — APPROVED** [review]

File: .github/workflows/build-images-docs-builder.yaml
Comment (learnitall) [inline .github/workflows/build-images-docs-builder.yaml]: Wouldn't this run the not trusted `update-docs-builder-image.sh` from the pull request branch, rather than the default trusted branch? If so, a user could change this script and run whatever they wanted in the workflow.

**Review by learnitall — CHANGES_REQUESTED** [review]
Quick question in regards to the `update-docs-builder-image.sh` before we merge.

File: .github/workflows/build-images-docs-builder.yaml
Comment (joestringer) [reply .github/workflows/build-images-docs-builder.yaml]: Note that triggering the overall workflow is only done with @cilium/docs-structure approval due to `environment: docs-builder` plus the cilium/cilium configuration for that environment. So we are partially relying here on those reviewers doing due diligence on any PRs requesting to change build infrastructure.

It would probably be better practice to do some combination of the following:

- Log out of Quay after images are pushed
- Directly call docker from the workflow to wrap & run the docs script

Comment (learnitall) [reply .github/workflows/build-images-docs-builder.yaml]: Ah ok that makes sense. Should this be non-blocking? Those options sound good. I think we could also save a copy of the trusted script in the beginning of the job and use it later o?

Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml]: Both proposals look good, but I'm not sure how to implement them.

> * Log out of Quay after images are pushed

I don't know how to explicitly logout from Quay. The Action we use supports a `logout` parameter, already enabled by default, but logging out occurs only at the end of the job. One solution I see would be to split the current job into several ones, starting a new one after `Build docs-builder image` (which I think is the only step that requires the log in)?

> * Directly call docker from the workflow to wrap & run the docs script

Do you mean replacing the script invocation with the following?"

```console
$ docker run --rm -v $PWD:/cilium -w /cilium cilium/docs-builder \
    bash -c "git config --global --add safe.directory /cilium && /cilium/Documentation/update-docs-builder-image.sh 'quay.io/...'"
```

Comment (aanm) [reply .github/workflows/build-images-docs-builder.yaml]: > One solution I see would be to split the current job into several ones, starting a new one after Build docs-builder image (which I think is the only step that requires the log in)?

I would also be in favor of this approach.

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by qmonnet** [comment]
OK so I got it working at last - sorry for the wall of logs on this page.

- [Successful deployment](https://github.com/cilium/cilium/actions/runs/5797099612)
- [Follow-up deployment](https://github.com/cilium/cilium/actions/runs/5797146270) with jobs skipped, given that we did not need to regenerate the image on top of auto-committer's commit.

Changes include:

- We now have 4 jobs instead of 1. The drawback is that we need to approve the deployment twice, given that two of these jobs, running in a succession, use the `docs-builder` environment.

- I also run the script in the Docker image that we have just generated (because it has both bash and git). This Docker image is modified by the PR author, but then so can be the script in the first place, and my understanding is that whatever the image contains, it still cannot access the environment from inside the container.

<details><summary>Incremental diff</summary>

``` diff
diff --git a/.github/workflows/build-images-docs-builder.yaml b/.github/workflows/build-images-docs-builder.yaml
index 191f9d4c9657..b4157a621f0d 100644
--- a/.github/workflows/build-images-docs-builder.yaml
+++ b/.github/workflows/build-images-docs-builder.yaml
@@ -2,12 +2,13 @@ name: Docs-builder Image Build
 
 # Any change in triggers needs to be reflected in the concurrency group.
 on:
-  pull_request_target:
+  pull_request:
     types:
       - opened
       - synchronize
       - reopened
     paths:
+      - .github/workflows/build-images-docs-builder.yaml
       - Documentation/Dockerfile
       - Documentation/requirements.txt
 
@@ -21,9 +22,13 @@ concurrency:
 
 jobs:
   build-and-push:
-    timeout-minutes: 45
-    environment: docs-builder
+    name: Build and push image
     runs-on: ubuntu-20.04
+    timeout-minutes: 30
+    environment: docs-builder
+    outputs:
+      tag: ${{ steps.docs-builder-tag.outputs.tag }}
+      digest: ${{ steps.docker-build-docs-builder.outputs.digest }}
     steps:
       - name: Checkout default branch (trusted)
         uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
@@ -45,11 +50,6 @@ jobs:
           persist-credentials: false
           ref: ${{ github.event.pull_request.head.sha }}
 
-      - name: Set up git
-        run: |
-          git config user.name "Cilium Imagebot"
-          git config user.email "noreply@cilium.io"
-
       - name: Generate image tag for docs-builder
         id: docs-builder-tag
         run: |
@@ -72,11 +72,12 @@ jobs:
           registry: quay.io
           username: ${{ secrets.QUAY_DOCS_BUILDER_USERNAME }}
           password: ${{ secrets.QUAY_DOCS_BUILDER_PASSWORD }}
+          logout: true
 
       - name: Build docs-builder image
         if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         uses: docker/build-push-action@2eb1c1961a95fc15694676618e422e8ba1d63825 # v4.1.1
-        id: docker_build_docs_builder
+        id: docker-build-docs-builder
         with:
           provenance: false
           context: ./Documentation
@@ -85,32 +86,48 @@ jobs:
           tags: |
             quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}
 
-      - name: Retrieve image digest
-        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
-        shell: bash
+  # Use a separate job for the steps below, to ensure we're no longer logged
+  # into Quay.io.
+  update-pr:
+    name: Update Pull Request with new image reference
+    needs: build-and-push
+    if: needs.build-and-push.outputs.digest
+    runs-on: ubuntu-20.04
+    timeout-minutes: 10
+    environment: docs-builder
+    steps:
+      - name: Checkout default branch (trusted)
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+        with:
+          ref: ${{ github.event.repository.default_branch }}
+          persist-credentials: false
+
+      - name: Set environment variables
+        uses: ./.github/actions/set-env-variables
+
+      # Warning: since this is a privileged workflow, subsequent workflow job
+      # steps must take care not to execute untrusted code.
+      - name: Checkout pull request branch (NOT TRUSTED)
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+        with:
+          persist-credentials: false
+          ref: ${{ github.event.pull_request.head.sha }}
+
+      - name: Set up git
         run: |
-          mkdir -p image-digest/
-          echo "## docs-builder" > image-digest/docs-builder.txt
-          echo "" >> image-digest/docs-builder.txt
-          echo "\`quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}\`" >> image-digest/docs-builder.txt
-          echo "" >> image-digest/docs-builder.txt
-
-      - name: Upload artifact digests
-        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
-        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
-        with:
-          name: image-digest docs-builder
-          path: image-digest
-          retention-days: 1
+          git config user.name "Cilium Imagebot"
+          git config user.email "noreply@cilium.io"
 
       - name: Update docs-builder image reference in CI workflow
-        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         run: |
-          Documentation/update-docs-builder-image.sh "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ steps.docs-builder-tag.outputs.tag }}@${{ steps.docker_build_docs_builder.outputs.digest }}"
+          NEW_IMAGE="quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ needs.build-and-push.outputs.tag }}@${{ needs.build-and-push.outputs.digest }}"
+          # Run in Docker to prevent the script from accessing the environment.
+          docker run --rm -v $PWD:/cilium -w /cilium "${NEW_IMAGE}" \
+              bash -c "git config --global --add safe.directory /cilium && \
+                       /cilium/Documentation/update-docs-builder-image.sh ${NEW_IMAGE}"
           git commit -sam "ci: update docs-builder"
 
       - name: Get token
-        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         id: get_token
         uses: cilium/actions-app-token@61a6271ce92ba02f49bf81c755685d59fb25a59a # v0.21.1
         with:
@@ -120,15 +137,47 @@ jobs:
       - name: Push changes into PR
         env:
           REF: ${{ github.event.pull_request.head.ref }}
-        if: ${{ steps.docs-builder-tag-in-repositories.outputs.exists == 'false' }}
         run: |
           git diff HEAD^
           git push https://x-access-token:${{ steps.get_token.outputs.app_token }}@github.com/${{ env.QUAY_ORGANIZATION }}/cilium.git HEAD:"$REF"
 
+  retrieve-digest:
+    name: Retrieve and upload image digest
+    needs: build-and-push
+    if: needs.build-and-push.outputs.digest
+    runs-on: ubuntu-20.04
+    timeout-minutes: 10
+    steps:
+      - name: Checkout default branch (trusted)
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+        with:
+          ref: ${{ github.event.repository.default_branch }}
+          persist-credentials: false
+
+      - name: Set environment variables
+        uses: ./.github/actions/set-env-variables
+
+      - name: Retrieve image digest
+        shell: bash
+        run: |
+          NEW_IMAGE="quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/docs-builder:${{ needs.build-and-push.outputs.tag }}@${{ needs.build-and-push.outputs.digest }}"
+          mkdir -p image-digest/
+          echo "## docs-builder" > image-digest/docs-builder.txt
+          echo "" >> image-digest/docs-builder.txt
+          echo "\`${NEW_IMAGE}\`" >> image-digest/docs-builder.txt
+          echo "" >> image-digest/docs-builder.txt
+
+      - name: Upload artifact digests
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        with:
+          name: image-digest docs-builder
+          path: image-digest
+          retention-days: 1
+
   image-digests:
     name: Display digests
+    needs: retrieve-digest
     runs-on: ubuntu-20.04
-    needs: build-and-push
     steps:
       - name: Create image digest directory
         shell: bash
```
</details>

Given that we've come this far, it would probably make sense to split the `docs-builder` environment into two distinct ones, the first with the Quay.io login secrets, the other one with the auto-committer credentials, and to pass only the necessary secrets to each job. Cc @aanm

File: .github/workflows/build-images-docs-builder.yaml
Comment (learnitall) [inline .github/workflows/build-images-docs-builder.yaml:180]: Could this step be merged into the previous job?

**Review by learnitall — COMMENTED** [review]
Awesome! One quick question about the last job, but LGTM.

File: .github/workflows/build-images-docs-builder.yaml
Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml:180]: :thinking: Let me see, I didn't pay too much attention to that step, but it seems that in the workflow I copied from (base image generation), we needed a separate job to retrieve images from the different jobs spawned from the matrix; and in our case, we don't need that, given that we only build one image. Is this what you mean?

I'll give it a try

Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml:180]: OK, seems like [it's still working](https://github.com/cilium/cilium/actions/runs/5799571701/job/15719921971)

<details><summary>Incremental diff</summary>

``` diff
diff --git a/.github/workflows/build-images-docs-builder.yaml b/.github/workflows/build-images-docs-builder.yaml
index b4157a621f0d..fa48b23aa73e 100644
--- a/.github/workflows/build-images-docs-builder.yaml
+++ b/.github/workflows/build-images-docs-builder.yaml
@@ -141,8 +141,8 @@ jobs:
           git diff HEAD^
           git push https://x-access-token:${{ steps.get_token.outputs.app_token }}@github.com/${{ env.QUAY_ORGANIZATION }}/cilium.git HEAD:"$REF"
 
-  retrieve-digest:
-    name: Retrieve and upload image digest
+  image-digest:
+    name: Retrieve and display image digest
     needs: build-and-push
     if: needs.build-and-push.outputs.digest
     runs-on: ubuntu-20.04
@@ -174,21 +174,6 @@ jobs:
           path: image-digest
           retention-days: 1
 
-  image-digests:
-    name: Display digests
-    needs: retrieve-digest
-    runs-on: ubuntu-20.04
-    steps:
-      - name: Create image digest directory
-        shell: bash
-        run: |
-          mkdir -p image-digest/
-
-      - name: Download image digest
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
-        with:
-          path: image-digest/
-
       - name: Output image digest
         shell: bash
         run: |
```
</details>

Comment (learnitall) [reply .github/workflows/build-images-docs-builder.yaml:180]: Yep that's what I was getting at! The incremental diff looks great!

**Review by learnitall — APPROVED** [review]
LGTM! :partying_face:

File: .github/workflows/build-images-docs-builder.yaml
Comment (aanm) [inline .github/workflows/build-images-docs-builder.yaml]: If we are back to pull_request target there's no need tos split it by jobs since PRs opened from forks won't have access to secrets.

**Review by aanm — APPROVED** [review]

File: .github/workflows/build-images-docs-builder.yaml
Comment (qmonnet) [reply .github/workflows/build-images-docs-builder.yaml]: :warning: We're not! This is just a test commit so I could validate the workflow

**Review by aanm — APPROVED** [review]

**Pull request merged**
